#summary Description of parameters to tune Kojak performance.
#labels Featured,Phase-Deploy,Parameters

= Navigation =
 * [TableOfContents Return to Table of Contents]

= Introduction =

This page describes in more detail the parameters used to tune the performance of Kojak. Parameters are stored in an ASCII text configuration file. An example configuration file can be found [SampleConfigurationFile here].

= Details =

|| *Parameter* || *Value* || *Recommended Range* || *Description* ||
|| cross_link || Number, Number, Number || see description || Specifies the sites of cross-linking and mass modification. Three values follow this parameter, with the format *chem1 chem2 mass*. Chem codes are integers signifying the reactive groups on the cross-linker: 1=amine reactive, 2=carboxyl reactive, 3=sulfhydryl reactive. The mass can be any real number, positive or negative.||
|| database || Text || n/a || The name of the fasta file containing the amino acid protein sequences (and decoy sequences) to be searched. It is recommended to include the path in the name of the file. ||
|| decoy_filter || Text || n/a || A short, case-sensitive string of characters that appears in the name of every decoy protein sequence in the database. The default value is the word _random_ ||
|| enrichment || Number || 0.0-1.0 || Value between 0 and 1 to describe 18O atom percent excess (APE). For example, 0.25 equals 25 APE. Set to 0 to disable enrichment. ||
|| fixed_modification || Text and Number || A-Y 0.1-5000.0 || Specify an amino acid character and modification mass separated by white space. For example: C 57.02146 will add that mass to all occurrences of cysteine in the sequence database. If multiple fixed modification masses are desired, use multiple fixed_modification declarations. Use @ to signify a modification to the N-terminus. ||
|| instrument || Number || 0 or 1 || Code to describe the instrument that was used to collect the spectra. 0 = Orbitrap, 1 = FTICR. ||
|| max_miscleavages || Number || 1-10 || Number of missed K or R cleavages allowed. Must be at least 2 to identify loop-links. ||
|| max_peptide_mass || Number || 1.0-50000.0 || Maximum peptide mass allowed when parsing the protein sequence database. ||
|| max_spectrum_peaks || Number || 1-50000 || Maximum number of MS2 peaks to analyze if using spectrum_processing. Peaks are kept in order of intensity, starting with the most intense. ||
|| min_peptide_mass || Number || 1.0-50000.0 || Minimum peptide mass allowed when parsing the protein sequence database. ||
|| modification || Text and Number || A-Y 0.1-5000.0 || Specify an amino acid character and variable modification mass separated by white space. For example: M 15.9949 will search methionine residues *with and without* this modification mass. If multiple variable modification masses are desired, use multiple fixed_modification declarations. Multiple variable modification masses can be applied to the same residue. Use @ to signify a modification to the N-terminus. ||
|| MS_data_file || Text || n/a || Name of the data file to be analyzed. Accepted formats are mzXML, mzML, Thermo RAW (on Windows with vendor drivers installed). It is recommended to include the path in the name of the file. ||
|| MS1_centroid || Number || 0 or 1 || Describes how peaks are stored in the precursor (MS1) scans. Set to 0 if peaks are profile. Set to 1 if peaks are centroided. ||
|| MS1_resolution || Number || 0-500000 || Describes the estimated resolution of precursor scan (MS1) peaks at 400 m/z. ||
|| MS2_centroid || Number || 0 or 1 || Describes how peaks are stored in the fragmentation (MS2 or MS/MS) scans. Set to 0 if peaks are profile. Set to 1 if peaks are centroided. ||
|| MS2_resolution || Number || 0-500000 || Describes the estimated resolution of fragmentation scan (MS2 or MS/MS) peaks at 400 m/z. ||
|| mono_link_mass || Number || 1.0-5000.0 || Additional mass added to single peptides during the database search. If data contain multiple mono-link masses, use multiple mono_link_mass lines in the configuration file, one per mass.||
|| output_file || Text || n/a || Name of the file for Kojak non-validated results. It is recommended to include the path in the name of the file. ||
|| percolator_file || Text || n/a || Name of the file for Kojak results that are formatted for validation with [Percolator]. It is recommended to include the path in the name of the file. ||
|| percolator_version || Number || n/a || Version of [Percolator] you will be using. Different versions of Percolator have different input requirements. Kojak will determine those for you based on this supplied number. ||
|| ppm_tolerance_frag || Number || 0.1-5000.0 || Tolerance used when matching peptide sequences to observed MS2 fragment ions. Unit is parts-per-million (PPM). ||
|| ppm_tolerance_pre || Number || 0.1-500.0 || Tolerance used when determining which peptides to search for a given MS2 spectrum based on its precursor ion mass. Unit is parts-per-million (PPM). ||
|| relaxed_analysis || Number || 0 or 1 || If set to 1 turns on *Relaxed Analysis Mode*. Set to 0 to use *Full Search Mode*. These modes are described elsewhere. It is recommended to use Relaxed Analysis Mode for all searches with a database size of 100 or more sequences. ||
|| search_dimers || Number || 0 or 1 || If set to 1 turns on searches for non-covalently linked dimers (not cross-links). This will no worse than double search time and is only recommended for highly concentrated and purified samples. ||
|| spectrum_processing || Number || 0 or 1 || If set to 1 turns on MS2 spectrum processing algorithm. This algorithm will collapse isotope distributions to the precursor peak and reduce the number of peaks to analyze to the number specified with the max_spectrum_peaks parameter. ||
|| top_count || Number || 1 to 10000 || This parameter is used for *Relaxed Analysis Mode* to specify the number of individual peptide scores to store in the first pass of the analysis. Setting below 200 is not recommended. Setting too high will unnecessarily consume computer resources. The recommended value is 500. ||
|| use_comet_xcorr || Number || 0 or 1 || If set to 1 will use the XCorr scoring function from the [http://comet-ms.sourceforge.net/ Comet] database search algorithm. If set to 0, will default to using Kojak's own rudimentary scoring function. The recommended value is 1. ||