#summary Description of parameters to tune Kojak performance.
#labels Featured,Phase-Deploy,Parameters

= Introduction =

Add your content here.


= Details =

|| *Parameter* || *Value* || *Range* || *Description* ||
|| cross_link_mass || Number || 1.0-5000.0 || Additional mass added to two cross-linked peptides during the database search. If data contain multiple cross-linkers, use multiple cross_link_mass lines in the configuration file, one per mass.||
|| database || Text || n/a || The name of the fasta file containing the amino acid protein sequences (and decoy sequences) to be searched. It is recommended to include the path in the name of the file. ||
|| decoy_filter || Text || n/a || A short, case-sensitive string of characters that appears in the name of every decoy protein sequence in the database. The default value is the word _random_ ||
|| enrichment || Number || 0.0-1.0 || Value between 0 and 1 to describe 18O atom percent excess (APE). For example, 0.25 equals 25 APE. Set to 0 to disable enrichment. ||
|| instrument || Number || 0 or 1 || Code to describe the instrument that was used to collect the spectra. 0 = Orbitrap, 1 = FTICR. ||
|| MS_data_file || Text || n/a || Name of the data file to be analyzed. Accepted formats are mzXML, mzML, Thermo RAW (on Windows with vendor drivers installed). It is recommended to include the path in the name of the file. ||